applications/tari_app_grpc/src/authentication/basic_auth.rs

```diff
@@ -0,0 +1,181 @@
+use tari_utilities::SafePassword;
+use zeroize::{Zeroize, Zeroizing};
+
+/// Implements [RFC 2617](https://www.ietf.org/rfc/rfc2617.txt#:~:text=The%20%22basic%22%20authentication%20scheme%20is,other%20realms%20on%20that%20server.)
+/// Represents the username and password contained within a Authenticate header.
+#[derive(Debug)]
+pub struct BasicAuthCredentials {
+    pub user_name: String,
+    pub password: SafePassword,
+}
+
+impl BasicAuthCredentials {
+    pub fn new(user_name: String, password: SafePassword) -> Self {
+        Self { user_name, password }
+    }
+
+    /// Creates a `Credentials` instance from a base64 `String`
+    /// which must encode user credentials as `username:password`
+    pub fn decode(auth_header_value: &str) -> Result<Self, BasicAuthError> {
+        let decoded = base64::decode(auth_header_value)?;
+        let as_utf8 = Zeroizing::new(String::from_utf8(decoded)?);
+
+        if let Some((user_name, password)) = as_utf8.split_once(':') {
+            let credentials = Self::new(user_name.into(), password.to_string().into());
+            return Ok(credentials);
+        }
+
+        Err(BasicAuthError::InvalidAuthorizationHeader)
+    }
+
+    /// Creates a `Credentials` instance from an HTTP Authorization header
+    /// which schema is a valid `Basic` HTTP Authorization Schema.
+    pub fn from_header(auth_header: &str) -> Result<BasicAuthCredentials, BasicAuthError> {
+        // check if its a valid basic auth header
+        let (auth_type, encoded_credentials) = auth_header
+            .split_once(' ')
+            .ok_or(BasicAuthError::InvalidAuthorizationHeader)?;
+
+        if encoded_credentials.contains(' ') {
+            // Invalid authorization token received
+            return Err(BasicAuthError::InvalidAuthorizationHeader);
+        }
+
+        // Check the provided authorization header
+        // to be a "Basic" authorization header
+        if auth_type.to_lowercase() != "basic" {
+            return Err(BasicAuthError::InvalidScheme(auth_type.to_string()));
+        }
+
+        let credentials = BasicAuthCredentials::decode(encoded_credentials)?;
+        Ok(credentials)
+    }
+
+    pub fn validate(&self, username: &str, password: &[u8]) -> Result<(), BasicAuthError> {
+        if self.user_name.as_bytes() != username.as_bytes() {
+            return Err(BasicAuthError::InvalidCredentials);
+        }
+        // These bytes can leak if the password is not utf-8, but since argon encoding is utf-8 the given
+        // password must be incorrect if conversion to utf-8 fails.
+        let bytes = self.password.reveal().to_vec();
+        let str_password = Zeroizing::new(String::from_utf8(bytes)?);
+        let header_password =
+            PasswordHash::parse(&str_password, Encoding::B64).map_err(|_| BasicAuthError::InvalidCredentials)?;
+        Argon2::default()
+            .verify_password(password, &header_password)
+            .map_err(|_| BasicAuthError::InvalidCredentials)?;
```

Comment:
Having a specific error for failed `Argon2` verification could be helpful  ?

Reply:
Will do!